src/server.rs

```diff
@@ -375,4 +380,83 @@ mod tests {
         subway_server.handle.stop().unwrap();
         upstream_dummy_server_handle.stop().unwrap();
     }
+
+    #[tokio::test]
+    async fn batch_requests_works() {
+        let (endpoint, upstream_dummy_server_handle) = upstream_dummy_server("127.0.0.1:9957").await;
+
+        // Server with max batch size 3
+        let subway_server = subway_server(endpoint, 9946, None, Some(3)).await;
+        let url = format!("ws://{}", subway_server.addr);
+        let client = ws_client(&url).await;
+
+        // Sending 3 request in a batch
+        let mut batch = BatchRequestBuilder::new();
+        batch.insert(PHO, rpc_params!()).unwrap();
+        batch.insert(PHO, rpc_params!()).unwrap();
+        batch.insert(PHO, rpc_params!()).unwrap();
+
+        let res = client.batch_request::<String>(batch).await.unwrap();
+        assert_eq!(res.num_successful_calls(), 3);
+
+        upstream_dummy_server_handle.stop().unwrap();
+    }
+
+    #[tokio::test]
+    async fn batch_requests_exceeds_max_size_errors() {
+        let (endpoint, upstream_dummy_server_handle) = upstream_dummy_server("127.0.0.1:9958").await;
+
+        // Server with max batch size 3
+        let subway_server = subway_server(endpoint, 9947, None, Some(3)).await;
+        let url = format!("ws://{}", subway_server.addr);
+        let client = ws_client(&url).await;
+
+        // Sending 4 request in a batch
+        let mut batch = BatchRequestBuilder::new();
+        batch.insert(PHO, rpc_params!()).unwrap();
+        batch.insert(PHO, rpc_params!()).unwrap();
+        batch.insert(PHO, rpc_params!()).unwrap();
+        batch.insert(PHO, rpc_params!()).unwrap();
+
+        // Due to the limitation of jsonrpsee client implementation,
+        // we can't check the error message when response batch id is `null`.
+        // E.g.
+        // Raw response - `{"jsonrpc":"2.0","error":{"code":-32010,"message":"The batch request was too large","data":"Exceeded max limit of 3"},"id":null}`
```

Comment:
can we set the batch id to some value so we can check the error message?

Reply:
RPC response json has null as its id value when max batch limit exceeds, I think this is because this error is not for a particular request in a batch. (I think it's the same if reached to rate limit, response json id will have null)
And unfortunately, there's no way for us to set id to response from outside jsonrpsee module.